README.md
# apiDoc

apiDoc creates a documentation from API descriptions in your source code.

![validate](https://github.com/apidoc/apidoc/workflows/validate/badge.svg)
[![Dependency Status](https://david-dm.org/apidoc/apidoc.svg)](https://david-dm.org/apidoc/apidoc)
[![NPM version](https://badge.fury.io/js/apidoc.svg)](http://badge.fury.io/js/apidoc)
[![Join the chat at https://gitter.im/apidoc/talk](https://badges.gitter.im/apidoc/talk.svg)](https://gitter.im/apidoc/talk?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

### [BOX-ID](https://github.com/box-id) Fork Changes

- Updated elixir block parser to use the old format by wrapping the doc with `@apidocs """ ... """`; version: `0.54.1`
- Optimized template for a better search and oder by group title; version: `0.55.1`
- Added simple function to search by multiple elements with a +; version: `0.56.0`

### Documentation: [apidocjs.com](http://apidocjs.com)

### [Live DEMO](http://apidocjs.com/example/)

## Installation

```bash
$ npm install -g apidoc
```

## Usage

Add some apidoc comments anywhere in your source code:

```java
/**
 * @api {get} /user/:id Request User information
 * @apiName GetUser
 * @apiGroup User
 *
 * @apiParam {Number} id User's unique ID.
 *
 * @apiSuccess {String} firstname Firstname of the User.
 * @apiSuccess {String} lastname  Lastname of the User.
 */
```

Now generate the documentation from `src/` into `doc/`.

```bash
$ apidoc -i src/ -o doc/
```

This repository contains and `example` folder from which you can generate a very complete documentation on an example api endpoint. It also contains best practice hints (in the `footer.md` file).

```bash
$ git clone https://github.com/apidoc/apidoc && cd apidoc
$ npm install --prod
$ ./bin/apidoc -i example -o /tmp/doc
$ $BROWSER /tmp/doc
```

### Programmatic usage

You can generate the documentation programmatically:

```ts
import path from "path";
import { createDoc } from "apidoc";

const doc = createDoc({
  src: path.resolve(__dirname, "src"),
  dest: path.resolve(__dirname, "doc"), // can be omitted if dryRun is true
  // if you don't want to generate the output files:
  dryRun: true,
  // if you don't want to see any log output:
  silent: true,
});

if (typeof doc !== "boolean") {
  // Documentation was generated!
  console.log(doc.data); // the parsed api documentation object
  console.log(doc.project); // the project information
}
```

Install type definitions (see [@types/apidoc](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/apidoc/index.d.ts)):

```bash
$ npm install -D @types/apidoc
```

## Docker image

You can use apidoc in Docker like this:

```bash
# first build the image after cloning this repository
docker build -t apidoc/apidoc .
# run it
docker run --rm -v $(pwd):/home/node/apidoc apidoc/apidoc -o outputdir -i inputdir
```

## Supported programming languages

- **C#, Go, Dart, Java, JavaScript, PHP, Scala** (all DocStyle capable languages):

  ```javascript
  /**
   * This is a comment.
   */
  ```

- **Clojure**:

  ```clojure
  ;;;;
  ;; This is a comment.
  ;;;;
  ```

- **CoffeeScript**:

  ```coffeescript
  ###
  This is a comment.
  ###
  ```

- **Elixir**:

  ```elixir
  #{
  # This is a comment.
  #}
  ```

- **Erlang**:

  ```erlang
  %{
  % This is a comment.
  %}
  ```

- **Perl**

  ```perl
  #**
  # This is a comment.
  #*
  ```

  ```perl
  =pod
  This is a comment.
  =cut
  ```

- **Python**

  ```python
  """
  This is a comment.
  """
  ```

- **Ruby**

  ```ruby
  =begin
  This is a comment.
  =end
  ```

## Plugins (extend apiDoc)

apiDoc will auto include installed plugins.

- [apidoc-plugin-schema](https://github.com/willfarrell/apidoc-plugin-schema) Generates and inject apidoc elements from api schemas. `npm install apidoc-plugin-schema`

For details and an example on how to implement your own plugin, please view [apidoc-plugin-test](https://github.com/apidoc/apidoc-plugin-test).

## Support

Please [create a new issue](https://github.com/apidoc/apidoc/issues/new/choose) if you have a suggestion/question or if you found a problem/bug.

## Contributing

apiDoc is a collaborative project. Pull requests are welcome. Please see the [CONTRIBUTING](https://github.com/apidoc/apidoc/blob/master/CONTRIBUTING.md) file.

## Build tools

- [flask-apidoc](https://pypi.python.org/pypi/flask-apidoc/) `pip install flask-apidoc`
- [grunt-apidoc](https://github.com/apidoc/grunt-apidoc) `npm install grunt-apidoc`.
- [gapidoc (gulp)](https://github.com/techgaun/gulp-apidoc) `npm install gapidoc`.
- [webpack-apidoc](https://github.com/c0b41/webpack-apidoc) `npm install --save-dev webpack-apidoc`.

## Integration

- [Eclipse Java apiDoc editor templates](https://github.com/skhani/eclipse_java_apiDoc_templates)
- [Eclipse plugin](https://github.com/DWand/eclipse_pdt_apiDoc_editor_templates)
- [Microsoft WebAPI](https://github.com/chehabz/grunt-edge-apidoc-webapi-generator)
- [Sublime Text plugin](https://github.com/DWand/ST3_apiDocAutocompletion)

## Converter

- [apidoc-swagger](https://github.com/fsbahman/apidoc-swagger)
- [apidoc-swagger-3](https://github.com/amanoooo/apidoc-swagger-3)
- [gulp-apidoc-swagger](https://github.com/fsbahman/gulp-apidoc-swagger)
- [Docmaster](https://github.com/bonzzy/docmaster)
- [apidoc-markdown](https://github.com/rigwild/apidoc-markdown)


CHANGELOG.md
# apiDoc Changelog

### 0.56.1 (Fork by box-id)

- Template: fixed navigation sorting
- Template: fixed search clear
- Template: handle current version in search

### 0.56.0 (Fork by box-id)

- Template: Added simple function to search by multiple elements with a +

### 0.55.1 (Fork by box-id)

- Template: Optimized template for a better search and oder by group title

### 0.54.1 (Fork by box-id)

- Updated elixir block parser to use the old format by wrapping the doc with `@apidocs """ ... """`

### 0.53.1

- Add tsx to default extensions parsed. (fix #1290)
- Compare methods with lowercase strings (fix #1284)
- Upgrade some dependencies

### 0.53.0

- Fix issue with aloneDisplay and non-ascii characters (#1273 by @KujoMiva)

### 0.52.0

- Allow cn, zh_cn or zh for Chinese language option and display meaningful message in console on error. fix #1248
- Update example files to not display an error (#1210)
- Update dependencies

### 0.51.1

- Fix incorrect warnings in some conditions (PR #1206 by @BrokenSourceCode, fix #1194)

### 0.51.0

- Emit warnings if there are inconsistencies in the apidoc params (PR #1189 by Lucas Cimon)
- Dependencies update

### 0.50.5

- Use root package.json values if not in apidoc key by Nick Mitchell (#1164)
- Fix search requiring lower-case input (#1173 by Alexander Weiss)
- Add yarn berry compatibility (#1144 by 이호영)
- Fix template path resolve (#1175 by Kevin Bull)
- Update chinese locale (#1184 by Jun Ding)
- Dependencies update

### 0.50.4

#### FIXED

- Fix bug with loading the CSS file for single page mode (`--single`) (fix #1142)

### 0.50.3

#### FIXED

- Fix bug where header/footer files could not be found in some conditions (fix #1107)

#### ENHANCEMENTS

- Add showRequiredLabels option (default: false) (fix #1095)

### 0.50.2

#### FIXED

- fix checkbox styling (fix #1077 via #1079 by @esaracco)
- make sure success block appears after body/query params (fix #1100)
- fix content-type value (fix #1084 via #1085 by @esaracco)
- fix display of types for apiHeader and apiParam (#1083 by @esaracco)
- fix tests on windows (#1070 by @donmahallem)

#### ENHANCEMENTS

- improve build speed (by @rottmann)
- make the message about no config files found a warning
- look for package.json in current working directory for apidoc config (fix #1086)
- update chinese translation
- add json response highlighting (#1082 by @esaracco)

### 0.50.1

- fix issue with sampleUrl set to "true" (fix #1058 via #1062 by @esaracco)
- fix not seeing all versions below (fix #1049 via #1061 by @esaracco)
- fix apiParams input elements (#1059) by Benjamin Knorr
- Remove the jQueryAjaxSetup configuration option (was not working anyway) (fix #1040 via #1041 by @esaracco)
- Add more files/folders to .npmignore
- Improve/fix group sorting (fix #1042 via #1044 by @esaracco)

### 0.50.0

#### BREAKING CHANGES

- --silent renamed to --quiet or -q for short
- --color renamed to --no-color to disable color in log output
- --parse was deleted because it is the same as --simulate
- --simulate renamed to --dry-run / -n
- Default output is not json, use "--log-format json" to get json
- --config / -c must be a path to a file, not a directory

#### FIXED

- Fix sampleUrl config option behavior:

  - if it is absent or `false`, no sample block will be generated
  - if it is `true`, the current page location will be used
  - if it is an url, that will be used
    Note that the option useHostUrlAsSampleUrl has no longer any effect and can be safely removed. Use a value of `true` for `sampleUrl` to get the same behavior.

- Fix non working sortFields function (fix #968)
- Fix syntax highlighting for markdown parsed text
- Fix color in logs was not showing since upgrade to winston 3
- Fix newlines being removed in pre tags (fix #809 by @JayBerlin)
- Fix header/footer in aloneDisplay mode (fix #905 via #1027 by @esaracco)
- Fix duplicated content in aloneDisplay mode (fix #938 via #1025 by @esaracco)
- Fix apiSampleRequest not working (fix #984 via #1023 by @esaracco)
- Fix error with parenthesis in apiGroup (fix #994 via #1022 by @esaracco)
- Fix absent diff on different api route (fix #963 via #1030 by @esaracco)

#### ADDED

- Add request method to compare template (#986 by @klickagent)
- Add some shorthand options, see apidoc -h
- Add --log-format option for json output (default is not json)
- Add ascii art in verbose mode
- Add --warn-error to treat warnings as error (fix #703)

#### ENHANCEMENTS

- HUGE CLEANUP OF THE CODEBASE
- Merge apidoc-core back into the apidoc repository. The apidoc-core repository is now read-only.
- Add .eslintrc.js and enforce coding style everywhere
- Update many dependencies to their latest version
- Make sidenav wider and template more responsive (fix #851)
- Big CSS code overhaul and change
- Ignore node_modules by default (fix #936)
- Add ASCII art in verbose mode
- Allow setting input and output in apidoc.json (fix #630)
- Take standard parameters of URL query string into account (#1014 by @esaracco)
- Allow colon character in apiDefine (fix #877 via #1021 by @esaracco)

### 0.29.0

- Fix broken handlebar
- [DEV] add a npm run serve command
- Remove space that was added for optional params
- Make the 'optional' label unselectable (#970 by Thomas Bruun)

### 0.28.1

- Fix issue with handlebar (#967)

### 0.28.0

- Use npm lockfile v2
- Update handlebar to fix vulnerability (#962)
- Fix runtime error in single page template (#944 via #945 by Ashlanfox)
- Implement apiBody parsing into multi page template (#953 by Seth Wheeler)

### 0.27.0

- Maintain comment order for @apiParam and @apiSuccess (fix #206 via #951 by Matheus Giovani)
- Add option for JavaScript output mode (#933 by refineddigital)
- Add new configuration variable `useHostUrlAsSampleUrl` (#930) by Shivam Kumar Jha
- Add new favicon (#926 by Francisko Arenas)
- Implement apiBody parsing into single page template (#916 by Seth Wheeler)

### 0.26.0

- Clean up useless CSS rule (#923)

elixir_example/input/test_controller_apidoc.ex
defmodule BXApiExampleWeb.TestController do
  use BXApiExampleWeb, :controller

  alias BXApiExample.Tests
  alias BXApiExample.Tests.Test

  @apidoc """
  @api {get} /test_entity/ List
  @apiName ListTests
  @apiGroup TestEntities
  @apiVersion 0.8.15

  @apiParam (Query Parameters) {Number} limit=500 Limit the number of returned elements
  @apiParam (Query Parameters) {String } [id] get a single id
  @apiParam (Query Parameters) {String[]} [id] get a list of id's

  @apiSuccessExample Success-Response:
    HTTP/1.1 200 OK
    [
      {
        "name": "API",
        "id": "5fcc2352-aa5a-4ad2-8c98-f6af5f9b7637",
        "tenant_id": "test"
      },
      {
        "name": "Tester",
        "id": "c59d0435-e0d8-4cb1-a8d2-16817fea405a",
        "tenant_id": "test"
      }
    ]

    @apiExample {curl} Example usage:
      curl -i http://api.example.com/api/test_entity?tenant_id=test&limit=2
  """
  def index(conn, params) do
    result = Tests.list(params, conn)
    render(conn, "index.json", tests: result)
  end

  @apidoc """
  @api {post} /test_entity/ Create
  @apiGroup TestEntities
  @apiName CreateTest
  @apiVersion 0.8.15

  @apiParam (Body) {String} name the name of the element
  @apiParam (Body) {String} tenant_id the tenant this element should belong to
  @apiParam (Body) {Object} settings={} a object of special settings for this element

  @apiSuccessExample Success-Response:
    HTTP/1.1 201 Created
    {
      "id": "5fcc2352-aa5a-4ad2-8c98-f6af5f9b7637",
      "name": "Test 183",
      "tenant_id": "test",
      "settings": { "foo": 1337 }
    }

  @apiExample {curl} Curl:
     curl -X POST -i http://api.example.com/api/test_entity -d '{
      "name": "Test 183",
      "tenant_id": "test",
      "settings": { "foo": 1337 }
    }'
  }
  """
  def create(conn, test_params) do
    with {:ok, %Test{} = result} <- Tests.create_Test(test_params) do
      conn
      |> put_status(:created)
      |> render("show.json", test: result)
    end
  end

  @apidoc """
  @api {post} /example_entity/ Create
  @apiGroup ExampleEntities
  @apiName CreateTest
  @apiVersion 0.8.15

  @apiParam (Body) {String} name the name of the element
  @apiParam (Body) {String} tenant_id the tenant this element should belong to
  @apiParam (Body) {Object} settings={} a object of special settings for this element

  @apiSuccessExample Success-Response:
    HTTP/1.1 201 Created
    {
      "id": "5fcc2352-aa5a-4ad2-8c98-f6af5f9b7637",
      "name": "Test 183",
      "tenant_id": "test",
      "settings": { "foo": 1337 }
    }

  @apiExample {curl} Curl:
     curl -X POST -i http://api.example.com/api/example_entity -d '{
      "name": "Test 183",
      "tenant_id": "test",
      "settings": { "foo": 1337 }
    }'
  }
  """
  def create(conn, test_params) do
    with {:ok, %Test{} = result} <- Tests.create_Test(test_params) do
      conn
      |> put_status(:created)
      |> render("show.json", test: result)
    end
  end

  @apidoc """
  @apiDefine MQTTExampleEntitiesData
  @apiExample {mqtt} Example message:
  {
    "id": "5fcc2352-aa5a-4ad2-8c98-f6af5f9b7637",
    "name": "Test 183",
    "tenant_id": "test",
    "settings": { "foo": 1337 }
  }
  """

  @apidoc """
  @api {mqtt} tenant/:tenant_id/example_entity/create ExampleEntity Created
  @apiGroup #MQTT-Messages
  @apiName MQTT ExampleEntities Created

  @apiVersion 0.8.15
  @apiDescription MQTT Message after a new example entity was created

  @apiUse MQTTExampleEntitiesData
  """
  defp get_topic(%{tenant_id: cid, id: _id}, "create") do
    ["tenant/#{cid}/example_entity/create"]
  end
end


elixir_example/input/test_controller_comment.ex
defmodule BXApiExampleWeb.TestController do
  use BXApiExampleWeb, :controller

  alias BXApiExample.Tests
  alias BXApiExample.Tests.Test

  # {
  # @api {get} /test_entity/ List Tests
  # @apiName ListTests
  # @apiGroup Tests
  # @apiVersion 0.31.0

  # @apiParam (Query Parameters) {Number} limit=500 Limit the number of returned elements
  # @apiParam (Query Parameters) {String } [id] get a single id
  # @apiParam (Query Parameters) {String[]} [id] get a list of id's

  # @apiSuccessExample Success-Response:
  #   HTTP/1.1 200 OK
  #   [
  #     {
  #       "name": "API",
  #       "id": "5fcc2352-aa5a-4ad2-8c98-f6af5f9b7637",
  #       "tenant_id": "test"
  #     },
  #     {
  #       "name": "Tester",
  #       "id": "c59d0435-e0d8-4cb1-a8d2-16817fea405a",
  #       "tenant_id": "test"
  #     }
  #   ]

  #   @apiExample {curl} Example usage:
  #     curl -i http://api.example.com/api/test_entity?tenant_id=test&limit=2
  # }
  def index(conn, params) do
    result = Tests.list(params, conn)
    render(conn, "index.json", tests: result)
  end

  # {
  # @api {post} /test_entity/ Create
  # @apiGroup Tests
  # @apiName CreateTest
  # @apiVersion 0.31.0

  # @apiParam (Body) {String} name the name of the element
  # @apiParam (Body) {String} tenant_id the tenant this element should belong to
  # @apiParam (Body) {Object} settings={} a object of special settings for this element

  # @apiSuccessExample Success-Response:
  #   HTTP/1.1 201 Created
  #   {
  #     "id": "5fcc2352-aa5a-4ad2-8c98-f6af5f9b7637",
  #     "name": "Test 183",
  #     "tenant_id": "test",
  #     "settings": { "foo": 1337 }
  #   }

  # @apiExample {curl} Curl:
  #    curl -X POST -i http://api.example.com/api/test_entity -d '{
  #     "name": "Test 183",
  #     "tenant_id": "test",
  #     "settings": { "foo": 1337 }
  #   }'
  # }
  def create(conn, test_params) do
    with {:ok, %Test{} = result} <- Tests.create_Test(test_params) do
      conn
      |> put_status(:created)
      |> render("show.json", test: result)
    end
  end
end


elixir_example/input/account_controller_2.ex
defmodule BXApiExampleWeb.TestController do
  use BXApiExampleWeb, :controller

  alias BXApiExample.Tests
  alias BXApiExample.Tests.Test

  @apidoc """
  @api {get} /test_entity/ List Tests
  @apiName ListTests
  @apiGroup Tests
  @apiVersion 0.31.0

  @apiParam (Query Parameters) {Number} limit=500 Limit the number of returned elements
  @apiParam (Query Parameters) {String } [id] get a single id
  @apiParam (Query Parameters) {String[]} [id] get a list of id's

  @apiSuccessExample Success-Response:
    HTTP/1.1 200 OK
    [
      {
        "name": "API",
        "id": "5fcc2352-aa5a-4ad2-8c98-f6af5f9b7637",
        "tenant_id": "test"
      },
      {
        "name": "Tester",
        "id": "c59d0435-e0d8-4cb1-a8d2-16817fea405a",
        "tenant_id": "test"
      }
    ]

    @apiExample {curl} Example usage:
      curl -i http://api.example.com/api/test_entity?tenant_id=test&limit=2
  """
  def index(conn, params) do
    result = Tests.list(params, conn)
    render(conn, "index.json", tests: result)
  end

  @apidoc """
  @api {post} /test_entity/ Create
  @apiGroup Tests
  @apiName CreateTest
  @apiVersion 0.31.0

  @apiParam (Body) {String} name the name of the element
  @apiParam (Body) {String} tenant_id the tenant this element should belong to
  @apiParam (Body) {Object} settings={} a object of special settings for this element

  @apiSuccessExample Success-Response:
    HTTP/1.1 201 Created
    {
      "id": "5fcc2352-aa5a-4ad2-8c98-f6af5f9b7637",
      "name": "Test 183",
      "tenant_id": "test",
      "settings": { "foo": 1337 }
    }

  @apiExample {curl} Curl:
     curl -X POST -i http://api.example.com/api/test_entity -d '{
      "name": "Test 183",
      "tenant_id": "test",
      "settings": { "foo": 1337 }
    }'
  }
  """
  def create(conn, test_params) do
    with {:ok, %Test{} = result} <- Tests.create_Test(test_params) do
      conn
      |> put_status(:created)
      |> render("show.json", test: result)
    end
  end
end
